plasmapy/particles/_factory.py

```diff
@@ -92,26 +97,34 @@ def _make_custom_particle_with_real_charge_number(
         # Add tests if this function becomes part of public API
         raise InvalidParticleError("Cannot create CustomParticle.")
 
-    if Z > base_particle.atomic_number:
+    if Z is not None and Z > base_particle.atomic_number:
         raise ChargeError("The charge number cannot exceed the atomic number.")
 
     mass = base_particle.mass - m_e * Z
     return CustomParticle(mass=mass, Z=Z, symbol=symbol)
 
 
-_particle_constructors = (
+_particle_constructors: tuple[
+    type
+    | Callable[
+        [ParticleLike | u.Quantity | CustomParticle | Sequence[ParticleLike]],
+        object,
+    ],
+    ...,
+] = (
     Particle,
     CustomParticle,
     CustomParticle._from_quantities,  # noqa: SLF001
     ParticleList,
     _make_custom_particle_with_real_charge_number,
 )
 
-_particle_types = (Particle, CustomParticle, ParticleList)
+_particle_types: tuple[type, ...] = (Particle, CustomParticle, ParticleList)
 
 
 def _physical_particle_factory(
-    *args, **kwargs
+    *args: ParticleLike | u.Quantity | CustomParticle | Sequence[ParticleLike],
```

Comment:
I thought about creating a type alias variable for all of this since it gets repeated, but if we make `_physical_particle_factory` into a public function, the type alias variable would need to also be public so that the documentation build doesn't break.